Add Machine Deletion Hooks Proposal (#862)
openshift/enhancements: enhancements/machine-api/machine-deletion-hooks.md

after a Machine has been marked for deletion, but prior to the Machine being
drained and/or associated instance terminated.

In particular, this will allow the Etcd Operator to prevent a Control Plane
Machine from being drained/removed until the Etcd Operator has had a chance
to synchronise the Etcd data to the replacement Machine.

This could also be used by other components such as Storage to allow them to
ensure safe detachment of volumes before the Machine is deleted.

### Goals

- Define an initial set of hook points for the deletion phase.
- Define an initial set and form of related annotations.
- Define basic expectations for a controller or process that responds to a
lifecycle hook.

### Non-Goals

- Create an exhaustive list of hooks; we can add more over time.
- Create new machine phases.
- Create a mechanism to signal what lifecycle point a machine is at currently.
- Dictate implementation of controllers that respond to the hooks.
- Implement ordering in the machine controller.
- Require anyone to use these hooks for normal machine operations, these are
strictly optional and for custom integrations only.

## Proposal

> - Utilize annotations to implement lifecycle hooks.

Author: I don't personally have concerns about using annotations, we discussed a number of options when implementing this upstream and it seemed appropriate and safe to use annotations. There are plenty of projects and APIs with Kubernetes that rely on annotations in some way or another, for example a lot of the service controller implementations across clouds add extra capability via annotations, so I think there's prior art here.

> What is the failure case if the user or X removed an expected annotation?

Not sure I follow this question? Are you asking what happens when a user removes one of the hook annotations, or if they mistakenly remove more annotations other than the one they intended?

> Today who has permissions to update machines is this only cluster-admin?

By default yes, cluster admins have access to the modify machines, plus anything/anyone that the admins delegate this responsibility too

> - Utilize annotations to implement lifecycle hooks.

Author: I think to a degree this is intended behaviour. For example in a scenario where there is no way to add new capacity (restricted environments/at quota for cloud), then having the ability for a user to manually intervene and allow the less safe approach will be required, even in the etcd case. I think we need to account for these on the hook implementing controller side - the hooks, like finalizers, should be considered a convenience but not a guarantee

> - Utilize annotations to implement lifecycle hooks.

Reviewer: Shouldn't the controller that needs the hook notice that there is a Machine without the annotation that is not being deleted and restore the missing settings?

> - Utilize annotations to implement lifecycle hooks.

Author: In the case that a user interferes and the Machine isn't being deleted, then yes, I would expect the hook implementing controller to re-add the hook and re-instate the protection.

What I think we need to avoid is if a user manually removes the hook while the Machine is being deleted, re-adding the hook at this point. 

To allow users to have an escape hatch when there are scenarios that require it, we need to make sure that the user can override the mechanism to allow the Machine API to continue with the removal even when the hook implementing controller deems it unsafe.

If the user removes it, and the hook implementing controller immediately re-adds it, then that could cause issues (and a game of hook ping pong) that then prevent the user being able to remove the Machine.

Perhaps having the escape hatch within the hooks themselves isn't the appropriate place, I guess the alternative is to have something which the implementing operator would observe and then remove the hook?

In the control plane case, we could have something on the CRD that we plan to implement that signals to the etcd operator the alternative cases that we need to cover (and therefore wouldn't need an escape hatch). I'm thinking something like rolloutStrategies inspired by those in statefulsets, defaulting to bursting, but also allowing "recreate"

> - Utilize annotations to implement lifecycle hooks.

Reviewer: If controllers do not add their hooks to resources where the delete timestamp is not nil (just like with finalizers), everything should work fine. The hook will be restored if it is removed during normal operation and it won't be restored if it is removed during deletion, so the user will have a way to bypass the hook.

> - Utilize annotations to implement lifecycle hooks.

Author: I've already captured this in the partner enhancement that talks about how Etcd will leverage this feature, I'll make sure it's included here in the risks/mitigations section

- Each lifecycle point can have 0 or more hooks.
- Hooks do not enforce ordering.
- Hooks found during machine reconciliation effectively pause reconciliation
until all hooks for that lifecycle point are removed from a machine's annotations.

### User Stories

#### Story 1
(pre-terminate) As an operator, I would like to have the ability to perform
different actions between the time a Machine is marked deleted in the API and

> the time the Machine is deleted from the infrastructure provider.

Author: Yep you've got it, I will try to clarify though.

With a pre-terminate, the Drain happens, we then pause with the Node empty (except DS and static pods), but still running. Once the hook is removed, we proceed to terminate the instance on the cloud provider side, which then causes the NodeLifecycle controller to remove the `Node` object from the cluster

> For example, when replacing a Control Plane Machine, ensure a new Control

Reviewer 2: For this story If etcd operator is gaining access to edit machines anyway why do we need a new thing? couldn't this just be a finaliser?

Author: I'll add a note about finalizers to the alternatives section, but for now, my view on this is that finalizers are for the garbage collector. We could teach Machine API to view other finalizers as a block and prevent it from doing any removal until the other finalizers are removed, but this goes against the assumptions that Kubernetes has around finalizers and could break things further down the line.

IMO, it's better to have a specific and well defined way of doing this for Machines than to cram extra logic into a field like finalizers that wasn't really intended for this purpose

Author: One thing that isn't listed in those quotes, which I was always told should be the case for finalizers, is that they shouldn't imply ordering, do we believe that's a true statement? Because using finalizers for this approach does imply ordering.

The benefit of the new system to me is that we have something generic that works with multiple use cases.

If we were to make this a finalizer specific thing, I don't think it's safe to make the Machine API wait until all other finalizers are gone, for example, this would cause a deadlock between PVC finalizers and the Machine API.

That means then that we have to teach Machine API about a specific etcd finalizer to wait on, which feels like we now don't have a clear separation of concerns between the two projects and very much makes the MAO and ETCDO tied together for the future.

Having something generic like this ensures MAO doesn't need to know anything about the etcd operator, and allows extension in the future for other projects. It makes the relationship one way, rather than bi-directional, which IMO is a cleaner implementation

Reviewer 2: Yeh that's the assumption for interoperability and the way I see this it would still be satisfied. To me our use case is the analogous to the one in the docs.
```
A common example of a finalizer is kubernetes.io/pv-protection, which prevents accidental deletion of PersistentVolume objects. When a PersistentVolume object is in use by a Pod, Kubernetes adds the pv-protection finalizer. If you try to delete the PersistentVolume, it enters a Terminating status, but the controller can't delete it because the finalizer exists. When the Pod stops using the PersistentVolume, Kubernetes clears the pv-protection finalizer, and the controller deletes the volume.
```

```
When a machine is master the etcd-protection finaliser is added. If you try to delete the Machine, it enters a Terminating status, but the controller can't delete it because the finalizer exists. When the Etcd stops using that Machine, the etcd-protection finalizer is cleared, and the controller deletes the machine.
```

The only order logic that we need is to enforce the etcd graceful removal is done before proceeding with the core infra deletion and can be implemented as above.
Isn't this otherwise actually proposing basically ordered finalisers but using an annotation instead? My concern is that's a loose public contract where any one can freely "petisize" any machine resulting in difficult to predict behaviour that will need to be maintained forever. Other than this I have no objections.

there are disruptions during replacement and we need the disk of the existing
instance to perform some disaster recovery operation.  This will also prevent
prolonged periods of having one fewer Control Plane host in the event the
replacement instance does not come up in a timely manner.

#### Story 2
(pre-drain) As an operator, I want the ability to utilize my own draining
controller instead of the logic built into the machine controller.  This will
allow me better flexibility and control over the lifecycle of workloads on each
node.

### Implementation Details/Notes/Constraints

For each defined lifecycle point, one or more hooks may be applied as an annotation
to the machine object. These annotations will pause reconciliation of a Machine object
until all hooks are resolved for that lifecycle point.
The hooks should be managed by a Hook Implementing Controller or other external
application, or manually created and removed by an administrator.

#### Lifecycle Points

##### pre-drain

`pre-drain.delete.hook.machine.cluster.x-k8s.io`

Hooks defined at this point will prevent the machine controller from draining a node

> after the machine object has been marked for deletion until the hooks are removed.

Author: > Again my concern about potentially blocking upgrades because of unexpected annotation it seems like this perhaps should have stricter validation enforcement via CR possibly?

These hooks only come into play when a machine is being deleted. This is not something that would normally happen during an upgrade scenario. Upgrades typically involve MCO which updates the components on the control plane nodes in place and reboots them for the changes to take effect, Machine API doesn't come into this process and neither would this feature.

I can't find another comment about blocking upgrades, have I missed something? 🤔 

> How will we know who added an annotation and why? The Reason for the hook might be very useful in understanding intent.

In the `Annotation Form` section below, we show some examples of what this might look like. The idea is that the name of the hook captures the reason for why the hook exists, and the value captures who created it. I'm hesitant to go down the route of a CR as it seems a lot of extra bulk for not a huge gain (I see the gain as a longer limit on the reason field).

I would expect for the etcd operator we would add the following hook:
```
pre-drain.delete.hook.machine.cluster-api.x-k8s.io/etcd-voting-member: openshift-etcd-operator
```
Would you want to put anything more into the context than this?

> after the machine object has been marked for deletion until the hooks are removed.

Reviewer 3: I only say block upgrade out of an error case. so for example for whatever reason, the hook was never deleted resulting in a block on drain. User for whatever reason does not know this is the case and attempts upgrade. In this case I believe this hook would/could block an upgrade.

> after the machine object has been marked for deletion until the hooks are removed.

Reviewer 3: Just a corner case we would want to provide clear signal on.

> after the machine object has been marked for deletion until the hooks are removed.

Author: > User for whatever reason does not know this is the case and attempts upgrade.

Machine API doesn't factor into upgrades so I don't think this would be a case that we need to worry about. During upgrades, MCO will rollout updates to the Nodes, which does its own draining and AFAIK isn't intended to hook into this

The only reason today that Machine API would be involved in an upgrade is if you need to create new capacity to allow the upgrade to go ahead, in which case these hooks wouldn't be involved either.

I think alerting on blocked removal of Machines should be sufficient signal for most cases given this, and we already have alerts like that set up

##### pre-terminate

`pre-terminate.delete.hook.machine.cluster.x-k8s.io`

Hooks defined at this point will prevent the machine-controller from
removing/terminating the instance in the cloud provider until the hooks are

> removed.

Reviewer 3: how will we mesh this with quorum-guard and or other PDB. basically if PDB is not met drain will not happen. So the expectation would be after hook is removed then the node will drain. We need to have clear signal.

...hook removed but PBD not met.

Author: I think we can represent this in a condition in the Machine status if we feel a need. Feedback from MAPI today isn't great, this is known and we are working on this.

I think today if a drain is blocked because of a PDB, we log a message but don't surface this onto a machine.

I could see us adding two conditions off the back of this proposal maybe, a terminable condition that reflects if pre-terminate hooks are present, and a drainable hook that reflects whether pre-drain hooks are present. For the drain situation itself, we could add a drained condition that could reflect any errors or blocks in the draining process when a Machine has a deletion timestamp

---
title: machine-deletion-hooks
authors:
  - "@JoelSpeed"
  - "@michaelgugino"
reviewers:
  - "@enxebre"
  - "@elmiko"
  - "@ademicev"
  - "@hexfusion"
approvers:

>   - TBD

Reviewer: I would expect someone from the cluster infrastructure team to be listed here. If you need a neutral moderator, let me know and I can help find someone.

Author: A neutral approver would be handy. I have already mentioned this to @deads2k as I was asking him some questions about finalizers, his expertise may make him an appropriate approver here

- Hooks do not enforce ordering.
- Hooks found during machine reconciliation effectively pause reconciliation
until all hooks for that lifecycle point are removed from a machine's annotations.

### User Stories

#### Story 1
(pre-terminate) As an operator, I would like to have the ability to perform
different actions between the time a Machine is marked deleted in the API and
the time the Machine is deleted from the infrastructure provider.

For example, when replacing a Control Plane Machine, ensure a new Control
Plane Machine has been successfully created and joined to the cluster before
removing the instance of the deleted Machine. This might be useful in case
there are disruptions during replacement and we need the disk of the existing
instance to perform some disaster recovery operation.  This will also prevent
prolonged periods of having one fewer Control Plane host in the event the
replacement instance does not come up in a timely manner.

#### Story 2
(pre-terminate) As an operator, I would to have the ability to perform some
action between the time that a Machine has been drained, but before it is
removed from the infrastructure provider.

> For example, when storage is attached to the Machine due to some pod placement,

Reviewer 2: This is generic enough as it should be built in. https://github.com/kubernetes-sigs/cluster-api/pull/4945

Author: Yes this is possibly not the best example, though was just copied across from the upstream proposal. I wonder if a better example might be something to do with GPUs and making sure they are detached properly, as this AFAIK is not a first party API field that can be observed in the same way that volume attachments can be.

CC @elmiko do you think GPUs might be a good example here?

Reviewer 4: interesting question, i haven't come across cases where _detaching_ the gpu has been an issue. most of the problems we see are around synchronizing the attachment/driver deployment for gpus. this might be worth looking into on the platforms that use post-creation gpu attachment.

Author: One use case idea I've seen in the past, when using an ELK/EFK stack, the logging daemon on the Node can become overwhelmed if particular applications are producing excessive logs. To ensure that none are lost when the node is removed, a logging operator may want to add a delay post drain to allow the logging daemon on the Node to catch up (ie once new log lines aren't being produced, it should be able to catch up). I'll add this as an extra example

Reviewer 4: that sounds like a solid example to me

an operator may want to check that the storage is detached before allowing the
removal of the machine from the infrastructure provider.

#### Story 3

> (pre-drain) As an operator, I want the ability to utilize my own draining

Reviewer 2: Do we know of a specific real need for this? I'd argue this can possibly be solved by exposing drain parameters?

Author: This possibly could be exposed by exposing drain parameters, though this is something we don't have today. I think this was also an example copied from the upstream enhancement so I'm not fully up to date on what the exact use case was for this.

Perhaps the idea is that this is generic enough that someone could utilize their own drain logic without CAPI/MAPI having to maintain the complexity of an exposed drain controller?

best time to take action.

For example, if an HIC manages a lifecycle hook at the pre-drain lifecycle-point,
then that controller should take action immediately after a Machine has a
DeletionTimestamp or enters the "Deleting" phase.

Fine-tuned coordination is not possible at this time; eg, it's not
possible to execute a pre-terminate hook only after a node has been drained.
This is reserved for future work.

##### Failure Mode

It is entirely up to the Hook Implementing Controller to determine when it is
prudent to remove a particular lifecycle hook. Some controllers may want to
'give up' after a certain time period, and others may want to block indefinitely.
Cluster operators should consider the characteristics of each controller before
utilizing them in their clusters.

### Future Work

#### Conditions

To ensure that users have visibility into why a Machine has not been removed yet,
we will add new conditions to the status of the Machine.

##### Drainable

The drainable condition will reflect whether, when deleted, the Machine would be
able to be drained. If any pre-drain hooks are present on the Machine, the
condition will be marked false.

##### Terminable

> The terminable condition will reflrect whether, when deleted, the Machine would

Reviewer: ```suggestion
The terminable condition will reflect whether, when deleted, the Machine would
```

## Drawbacks

* If misused, this could prevent Machines from being deleted and lead to an increase in support
cases to determine if it is safe to override the feature.

## Alternatives

### Custom Machine Controller
Require advanced users to fork and customize.
This can already be done if someone chooses, so not much of a solution.

### Finalizers
We define additional finalizers, but this really only implies the deletion lifecycle point.
A misbehaving controller that accidentally removes finalizers could have undesirable effects.

The proposal today effectively extends the finalizer concept to allow two stages of finalization
(pre and post drain) in conjunction with the existing Machine finalizer.

If we were to leverage finalizers for this purpose, there are a few different ways this could work,
as set out below.

#### Machine API waits on all Finalizers

In this scenario, the Machine API would not start draining/removing a Machine until all finalizers
other than its own one have been removed. This would allow future controllers to manage the removal
of the Machine without having Machine API have to know anything about their specific implementation.

> However, there could be a scenario where another controller also needs to wait for other finalizers

Reviewer 2: Can we add a few examples of these scenarios here?
Multiple controllers operating over machines needing to wait for each others and so introducing state sounds like possibly not the best angle to look at the particular problem being solved and so supporting a design for this might be a flaw.

Since this is proposing to introduce a new public API can we describe at least 2 valid scenarios/use cases that this would solve? By looking at user stories section only Story 1 seems a real need.

Author: I'm not sure I can think of a specific example right now, but having suggested this as a design for this problem, I think it's reasonable to think that there may be other cases of this in the future. To me, "waiting for everyone else to go first" is fragile and seems like the kind of thing where if we assume it's fine now, will backfire later.

If we looked at this a slightly different way and use Nodes as an example, on Nodes, if a PVC is used, there are PVC finalizers added that prevent the Node removal until the PVCs have been detached. If we blocked draining the Node waiting for other finalizers to go, this would create a deadlock as the PVC needs the pods to be removed before it can remove the finalizer it created.

I know that doesn't affect this specific situation but if we do go down this route, it would open us up to this kind of dependency in the future

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Glossary

### Lifecycle Hook
A specific point in a machine's reconciliation lifecycle where execution of
normal machine-controller behaviour is paused or modified.

### Deletion Phase
Describes when a machine has been marked for deletion but is still present
in the API.  Various actions happen during this phase, such as draining a node,
deleting an instance from a cloud provider, and deleting the node object.

### Hook Implementing Controller (HIC)
The Hook Implementing Controller describes a controller, other than the
machine controller, that adds, removes, and/or responds to a particular
lifecycle hook. Each lifecycle hook should have a single HIC, but an HIC
can optionally manage one or more hooks.

## Summary

Defines a set of annotations that can be applied to a Machine which affect the

> linear progress of a Machine’s lifecycle after a Machine has been marked for

Reviewer 5: What does "linear progress" mean in this sentence.  I can see ordering is a non-goal a little further down.

> deletion. These annotations are optional and may be applied during Machine

Reviewer 5: can they be modified after deletion and before api removal?

Reviewer 5: > Yes, they need to be modified because they have to be removed before the Machine can be deleted. In theory new ones could be added, I don't think this really affects us too much as we expect these will always be added before the machine is deleted, if new ones are added post delete, we will block again, but the result may not be what is intended (eg we already called for the cloud provider to terminate the instance and then someone adds a pre-drain).

I suggest getting specific here.  It seems reasonable to prevent the addition of new hooks after deletion has begun.

ensure the main workloads are removed and no longer adding to the log backlog
before waiting for the log exporter to catch up on the synchronisation process,
ensuring all application logs are captured.

#### Story 3
(pre-drain) As an operator, I want the ability to utilize my own draining
controller instead of the logic built into the machine controller. This will
allow me better flexibility and control over the lifecycle of workloads on each
node.

For example this would allow me to prioritise moving particular mission critical
applications first to ensure that service interruptions are minimised in cases
where cluster capacity is limited. Ordering is not supported today in drain
libraries and as such, this would require a custom drain provider.

### Implementation Details/Notes/Constraints

For each defined lifecycle point, one or more hooks may be applied as an annotation
to the machine object. These annotations will pause reconciliation of a Machine object
until all hooks are resolved for that lifecycle point.
The hooks should be managed by a Hook Implementing Controller or other external
application, or manually created and removed by an administrator.

> #### Lifecycle Points

Reviewer 5: the different lifecycle points are a potentially good reason to avoid finalizers (thank you for the alternatives section).  But if you have a structured shape, why is it better to use an annotation instead of creating a spec field indicating what you want to have happen at these points?

Reviewer 2: To give more context on hypershift a capi machine or any other capi resource is considered an internal impl detail. NodePool is the consume facing interface and this hook concept does not exist.
Also the only known real use case is control plane and etcd which does not apply to hypershift since the control plane run as pods.

##### Hook Implementing Controllers must

* Watch Machine objects and determine when an appropriate action must be taken.
* After completing the desired hook action, remove the hook annotation.

##### Hook Implementing Controllers may

* Watch Machine objects and add a hook annotation as desired by the cluster
administrator.
* Coordinate with other Hook Implementing Controllers through any means
possible, such as using common annotations, CRDs, etc. For example, one hook
controller could set an annotation indicating it has finished its work, and
another hook controller could wait for the presence of the annotation before
proceeding.

#### Determining when to take action

A Hook Implementing Controller should watch Machines and determine when is the
best time to take action.

For example, if an HIC manages a lifecycle hook at the pre-drain lifecycle-point,
then that controller should take action immediately after a Machine has a
DeletionTimestamp or enters the "Deleting" phase.

Fine-tuned coordination is not possible at this time; eg, it's not
possible to execute a pre-terminate hook only after a node has been drained.
This is reserved for future work.

##### Failure Mode

> It is entirely up to the Hook Implementing Controller to determine when it is

Reviewer 5: Is there any supported mechanism for a hook to report status of its work on the machine itself?

Reviewer 5: > At the moment no, will have to have a think about how we could provide that

I think this is important before shipping the feature.  Consider future-us supporting this API and some hook is suspected of perma-failing.  Insights wouldn't be able to provide us information about what is failing without a reporting mechanism, correct?

prudent to remove a particular lifecycle hook. Some controllers may want to
'give up' after a certain time period, and others may want to block indefinitely.
Cluster operators should consider the characteristics of each controller before
utilizing them in their clusters.

### Future Work

#### Conditions

> To ensure that users have visibility into why a Machine has not been removed yet,

Reviewer 5: Visibility seems logically required before we can support this feature in a release.  Why was this work deferred as future?

#### Tech Preview -> GA

This feature will be released straight to GA.

#### Removing a deprecated feature

This proposal does not involve removing any deprecated features.

### Upgrade / Downgrade Strategy

As this feature is opt-in, there is no need to account for upgrades.

Any new Hook Implementing Controller will be able to add the hooks at
any point and they will be observed when the updated Machine controller
is deployed via the CVO.

On downgrades, HICs may wish to remove the annotations, but they are not
harmful if left behind.

### Version Skew Strategy

We do not expect any issues with version skew as this is a new feature.

## Implementation History

* This has already been implemented in Cluster API

## Drawbacks

* If misused, this could prevent Machines from being deleted and lead to an increase in support
cases to determine if it is safe to override the feature.

## Alternatives

### Custom Machine Controller
Require advanced users to fork and customize.
This can already be done if someone chooses, so not much of a solution.

### Finalizers
We define additional finalizers, but this really only implies the deletion lifecycle point.

> A misbehaving controller that accidentally removes finalizers could have undesirable effects.

Reviewer 5: doesn't the misbehaving bit also apply to the annotation?

Author: Updated this with a good example, if you remove too many finalizers, the Machine gets removed from etcd, possibly before the machine has been deleted on the cloud side, leaking the instance and adding cost. Removing too many finalizers wouldn't result in the machine instance being leaked.

The proposal today effectively extends the finalizer concept to allow two stages of finalization

> (pre and post drain) in conjunction with the existing Machine finalizer.

Reviewer 5: it appears that your annotation has a pattern for expressing: "wait at this stage".  Could that same pattern be used for finalizers?  If you did, does it resolve your, "what do I wait on" problem and simultaneously avoid needing to teach people about new API concepts?

of a particular finalizer it should wait on. This solution allows the third party operator to opt
in to having their cleanup logic block the removal of the the Machine, without having to have the
Machine controller aware of the finalizer ahead of time.

As this approach includes the use of annotations to prevent the Machine controller from taking any
action, it is very similar to the approach described in the proposal above.
Since any deletion hook annotation prevents the Mahcine controller from removing its own finalizer,
this secondary annotation + finalizer combination doesn't add much value over the existing proposal.
As it also means we deviate from the upstream solution, and may have to account for this in the
future with additional complexity, this is not a preferred solution.

### Status Field
Harder for users to modify or set hooks during machine creation.
How would a user remove a hook if a controller that is supposed to remove it is misbehaving?
We’d probably need an annotation like ‘skip-hook-xyz’ or similar and that seems redundant
to just using annotations in the first place.

### Spec Field

> We probably don’t want other controllers dynamically adding and removing spec fields on an object.

Reviewer 5: I can't tell if this is a concern about granting the permission to another controller or a claim that server-side-apply doesn't allow for "I want this field to exist in spec" (I think it does).

Reviewer 5: setting annotations requires spec update powers, so I don't think you're gaining anything rbac-wise.

Reviewer 5: > I'll expand this once I've double checked the upstream conversation that happened around this, but as I mentioned in another comment, I think this was more about trying to avoid mixing controller owned fields into the spec which is normally owned by a user. Will get back to this

I'm interested in seeing a link from here.  Given that server-side-apply is based on per-field ownership, a subdivided spec is tractable.

Author: Looking through the upstream conversation on the enhancement PR (I haven't found the original google doc yet): 

Part of the reason to use annotations is that the keys have to be unique, so it's less likely that there would be conflicts between different controllers who are dealing with the hooks. (https://github.com/kubernetes-sigs/cluster-api/pull/3132#discussion_r446360507)

Another comment specifically on this title is leading to a conversation about possibly making this a spec field in a future iteration, though I don't think anything has been actioned from that (https://github.com/kubernetes-sigs/cluster-api/pull/3132#discussion_r448618893)

Some more concern about the fact that using annotations is unversioned later in the thread, but agreement that this is ok for an alpha API. Interestingly, the CAPI project is looking to move to beta in the next few weeks, there has been no further conversation around this annotation API, I'll poke them to remind them. (https://github.com/kubernetes-sigs/cluster-api/pull/3132#issuecomment-661104241)

If we can convince the upstream community to revisit how this mechanism works and then it moves to a spec field, then that's great, I think that would work well for us, I'm just a little concerned about the timing here. I'd like to provide _something_ for the control plane scaling project to use, preferably with the same API as CAPI so that there's an easy conversion in the future. Perhaps we can implement the annotations temporarily and then move to a spec before we cut a release, assuming we can upstream consensus 🤔

Reviewer 5: This still isn't clear to me.  Ownership in the spec can be separated, it's what server-side-aply is for.  In addition, the permissions required to modify annotations are identical to permissions required to modify spec, so I don't see a benefit permissions-wise.

Duplicates are prevented by setting the appropriate list merge key and enforcing as we do for other fields.

Conflicts can be managed via server-side apply, patches, or (default) conditional updates.  Using a spec field increases expressivity.

Using an annotation for evaluation guarantees that there will be a migration later when a real field is developed.

been terminated.

To prevent potentially confusing behaviour, once a Machine has been marked for deletion,
it is expected that hook implementing controllers should not attempt to add new deletion
hooks. The hook implementing controller is expected to detect the deletion in progress
and perform any clean up logic necessary.

To prevent potential issues here, and to ensure users do not accidentally add hooks,
we will leverage the existing Machine API webhooks to prevent new annotations of the
deletion hook format from being added to Machines after they have been marked for
deletion.

##### Hook failure

The machine controller should not timeout or otherwise consider the lifecycle
hook as 'failed.'  Only the Hook Implementing Controller (or the end user in
extenuating circumstances) may decide to remove a particular lifecycle hook
to allow the machine controller to progress past the corresponding lifecycle-point.

The Machine status will contain conditions identifying why the Machine controller
has not progressed in removing the Machine. It is expected that hook implementing
controllers should signal to users when they are having issues via some mechanism

> (eg. their `ClusterOperator` status) and that the conditions on the Machine should

Reviewer 5: The clusteroperator status isn't likely to be an entry point given the cardinality differences between the hook provider and the number of machines.

Is there any mechanism to be included in the machine status.  These hooks are pretty deep and condition names can be scope limited.  What about encouraging these hooks to place a condition aligned to their name when they fail?

#### Hook Implementing Controller Design

Hook Implementing Controller is the component that manages a particular
lifecycle hook.

##### Hook Implementing Controllers must

* Watch Machine objects and determine when an appropriate action must be taken.
* After completing the desired hook action, remove the hook annotation.

##### Hook Implementing Controllers may

* Watch Machine objects and add a hook annotation as desired by the cluster
administrator.
* Coordinate with other Hook Implementing Controllers through any means
possible, such as using common annotations, CRDs, etc. For example, one hook
controller could set an annotation indicating it has finished its work, and
another hook controller could wait for the presence of the annotation before
proceeding.

#### Determining when to take action

A Hook Implementing Controller should watch Machines and determine when is the
best time to take action.

For example, if an HIC manages a lifecycle hook at the pre-drain lifecycle-point,
then that controller should take action immediately after a Machine has a
DeletionTimestamp or enters the "Deleting" phase.

Fine-tuned coordination is not possible at this time; eg, it's not

> possible to execute a pre-terminate hook only after a node has been drained.

Reviewer 5: This isn't obvious to me.  Why can a hook implementing controller not detect that the node has been drained?

This is reserved for future work.

##### Failure Mode

It is entirely up to the Hook Implementing Controller to determine when it is
prudent to remove a particular lifecycle hook. Some controllers may want to

> 'give up' after a certain time period, and others may want to block indefinitely.

Reviewer 5: This is an example of some of the expressivity that an actual spec field could provide.

authors:
  - "@JoelSpeed"
  - "@michaelgugino"
reviewers:
  - "@enxebre"
  - "@elmiko"
  - "@ademicev"
  - "@hexfusion"
approvers:
  - "@elmiko"
  - "@hexfusion"
  - "@deads2k"
creation-date: 2021-08-10
last-updated: 2021-11-11
status: implementable
see-also:
  - [Cluster API Upstream Equivalent](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20200602-machine-deletion-phase-hooks.md)
---

# Machine Deletion Hooks

## Release Signoff Checklist

- [x] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Operational readiness criteria is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Glossary

### Lifecycle Hook
A specific point in a machine's reconciliation lifecycle where execution of
normal machine-controller behaviour is paused or modified.

### Deletion Phase
Describes when a machine has been marked for deletion but is still present
in the API.  Various actions happen during this phase, such as draining a node,

> deleting an instance from a cloud provider, and deleting the node object.

Reviewer 4: minor nit for consistency (we seem to use "cloud provider" or "infrastructure provider" interchangeably)
```suggestion
deleting an instance from an infrastructure provider, and deleting the node object.
```

### Hook Implementing Controller (HIC)
The Hook Implementing Controller describes a controller, other than the
machine controller, that adds, removes, and/or responds to a particular
lifecycle hook. Each lifecycle hook should have a single HIC, but an HIC
can optionally manage one or more hooks.

## Summary

> Defines a set of lifecycle hookos that can be applied to a Machine which can be used

Reviewer 4: minor nit
```suggestion
Defines a set of lifecycle hooks that can be applied to a Machine which can be used
```

  // certain predefined points within the machine lifecycle
  // +optional
  lifecycleHooks LifecycleHooks `json:"lifecycleHooks,omitempty"`
}

type LifecycleHooks struct {
  // +optional
  PreDrain []LifecycleHook `json:"preDrain,omitempty"`

  // +optional
  PreTerminate []LifecycleHook `json:"preTerminate,omitempty"`
}

// LifecycleHook represents a single instance of a lifecycle hook
type LifecycleHook struct {
  // Name defines a unique name for the lifcycle hook.
  // The name should be unique and descriptive, ideally 1-3 words, in camelCase with no spaces.
  // Names must be unique and should only be managed by a single entity.
  // +kubebuilder:validation:Pattern:="[A-Za-z]+"
  // +required
  Name string `json:"name"`

  // Owner defines the owner of the lifecycle hook.
  // This should be descriptive enough so that users can identify
  // who/what is responsible for blocking the lifecycle.
  // This could be the name of a controller or an administrator
  // managing the hook.
  // +required
  Owner string `json:"owner"`
}
```

### Implementation Details/Notes/Constraints

For each defined lifecycle point, one or more hooks may be applied within the spec

> of the machine object. These annotations will pause reconciliation of a Machine

Reviewer 4: still referencing annotations here

object until all hooks are resolved for that lifecycle point.
The hooks should be managed by a Hook Implementing Controller or other external
application, or manually created and removed by an administrator.

#### Lifecycle Points

##### pre-drain

```yaml
lifecycleHooks:
  preDrain:
  - name: <hook-name>
    owner: <hook-owner>
```

Hooks defined at this point will prevent the machine controller from draining a node
after the machine object has been marked for deletion until the hooks are removed.

##### pre-terminate

```yaml
lifecycleHooks:
  preTerminate:
  - name: <hook-name>
    owner: <hook-owner>
```

Hooks defined at this point will prevent the machine-controller from

> removing/terminating the instance in the cloud provider until the hooks are

Reviewer 4: minor nit for consistency
```suggestion
removing/terminating the instance in the infrastructure provider until the hooks are
```

removed.

"pre-terminate" has been chosen over "pre-delete" because "terminate" is more

> easily associated with an instance being removed from the cloud or

Reviewer 4: i think we could just drop "cloud" here
```suggestion
easily associated with an instance being removed from the 
```

relatively easy for the end user to find.

##### Hook ordering

The machine controller will not attempt to enforce any ordering of hooks.
No ordering should be expected by the machine controller.

Hook Implementing Controllers may choose to provide a mechanism to allow
ordering amongst themselves via whatever means HICs determine.
Examples could be using CRDs external to the Machine API, gRPC communications,
or additional annotations on the Machine or other objects.

#### Hook Implementing Controller Design

Hook Implementing Controller is the component that manages a particular
lifecycle hook.

##### Hook Implementing Controllers must

* Watch Machine objects and determine when an appropriate action must be taken.
* After completing the desired hook action, remove the hook.

##### Hook Implementing Controllers may

* Watch Machine objects and add a hook as desired by the cluster administrator.
* Coordinate with other Hook Implementing Controllers through any means
possible, such as using common annotations, CRDs, etc. For example, one hook
controller could set an annotation indicating it has finished its work, and

> another hook controller could wait for the presence of the annotation before

Reviewer 4: still referencing annotations here

Author: I think this one is a legitimate reference to an annotation, WDYT?

Reviewer 4: yes, sorry, my confusion

other than its own one have been removed. This would allow future controllers to manage the removal
of the Machine without having Machine API have to know anything about their specific implementation.
However, there could be a scenario where another controller also needs to wait for other finalizers
to be removed, in such a case, unless the Machine API or this new controller are aware of each other
there would be a deadlock created.

Relying on all finalizers being removed before draining/removing could become fragile in the future
due to the above, and as such is not a preferred solution.

#### Machine API waits on some Finalizers

In this scenario, Machine API would have a list of known finalizers that would prevent it from taking
certain actions. For example a known etcd-quorum finalizer could prevent the Machine controller from
draining the Machine.

This creates a tight coupling between the Machine controller and other components within OpenShift.
The upstream solution today, matches the proposal set out in this document. As we are experimenting
with Cluster API in HyperShift, Central Machine Management and for future IPI clusters, it is

> preferred to not create a tight coupling with these components and to copy the upstream solution to

Reviewer 2: To clarify, none of these fields exists upstream today, though nothing preclude us from proposing this approach. A hook mechanism is to be discussed as part of https://github.com/kubernetes-sigs/cluster-api/discussions/5556#discussioncomment-1575309

removed by the HIC after the Machine is terminated, or once the hook is no
longer required.

```go
type MachineSpec struct {
  ... // Existing fields

  // lifecycleHooks allow users to pause operations on the machine at
  // certain predefined points within the machine lifecycle
  // +optional
  lifecycleHooks LifecycleHooks `json:"lifecycleHooks,omitempty"`
}

type LifecycleHooks struct {
  // +optional
  PreDrain []LifecycleHook `json:"preDrain,omitempty"`

  // +optional
  PreTerminate []LifecycleHook `json:"preTerminate,omitempty"`
}

// LifecycleHook represents a single instance of a lifecycle hook
type LifecycleHook struct {
  // Name defines a unique name for the lifcycle hook.
  // The name should be unique and descriptive, ideally 1-3 words, in camelCase with no spaces.
  // Names must be unique and should only be managed by a single entity.
  // +kubebuilder:validation:Pattern:="[A-Za-z]+"
  // +required
  Name string `json:"name"`

  // Owner defines the owner of the lifecycle hook.
  // This should be descriptive enough so that users can identify
  // who/what is responsible for blocking the lifecycle.

>   // This could be the name of a controller or an administrator

Reviewer 5: nit: example with `clusteroperator/etcd` perhaps?

  // managing the hook.
  // +required
  Owner string `json:"owner"`
}
```

> 

Reviewer 2: can we please flesh out the new conditions types and reasons here?

to allow the machine controller to progress past the corresponding lifecycle-point.

The Machine status will contain conditions identifying why the Machine controller
has not progressed in removing the Machine. This will be in the form of new `Drainable`
and `Terminable` conditions.

It is expected that hook implementing controllers should signal to users when they
are having complete system issues via some mechanism (eg. their `ClusterOperator` status if the issue is endangering
the cluster health) and that they should also use conditions on the Machine to
identify issues with removing the hook from a particular Machine.

When errors occur, hook implementing controllers are encouraged to add a condition
to Machines with details of the error. The condition `Type` should be formad as the
hook name, prefixed by the lifecycle point. For example, for a `preDrain` hook with name

> `MigrateImportantApp`, the condition `Type` would be `PreDrainMigrateImportantApp`.

Reviewer 5: `PreDrainMigrateImportantAppSucceeded` ?  So there's a clear meaning to true an false.
